cql3/values.hh

```diff
@@ -86,38 +96,67 @@ public:
         return std::holds_alternative<unset_value>(_data);
     }
     bool is_value() const {
-        return std::holds_alternative<fragmented_temporary_buffer::view>(_data);
-    }
-    std::optional<fragmented_temporary_buffer::view> data() const {
-        if (auto pdata = std::get_if<fragmented_temporary_buffer::view>(&_data)) {
-            return *pdata;
-        }
-        return {};
+        return _data.index() <= 1;
     }
     explicit operator bool() const {
         return is_value();
     }
-    const fragmented_temporary_buffer::view* operator->() const {
-        return &std::get<fragmented_temporary_buffer::view>(_data);
+
+    template <typename Func>
+    requires std::invocable<Func, const managed_bytes_view&> && std::invocable<Func, const fragmented_temporary_buffer::view&>
+    decltype(auto) with_value(Func f) const {
+        switch (_data.index()) {
+        case 0: return f(std::get<fragmented_temporary_buffer::view>(_data));
+        default: return f(std::get<managed_bytes_view>(_data));
+        }
     }
-    const fragmented_temporary_buffer::view& operator*() const {
-        return std::get<fragmented_temporary_buffer::view>(_data);
+
+    template <typename Func>
+    requires std::invocable<Func, bytes_view>
+    decltype(auto) with_linearized(Func f) const {
+        return with_value([&] (const FragmentedView auto& v) {
+            return ::with_linearized(v, std::forward<Func>(f));
+        });
     }
 
-    bool operator==(const raw_value_view& other) const {
-        if (_data.index() != other._data.index()) {
-            return false;
-        }
-        if (is_value() && **this != *other) {
-            return false;
-        }
-        return true;
+    size_t size_bytes() const {
+        return with_value([&] (const FragmentedView auto& v) {
+            return v.size_bytes();
+        });
+    }
+
+    template <typename ValueType>
+    ValueType deserialize(const abstract_type& t) const {
+        return value_cast<ValueType>(with_value([&] (const FragmentedView auto& v) { return t.deserialize(v); }));
+    }
+
+    template <typename ValueType>
+    ValueType deserialize(const collection_type_impl& t, cql_serialization_format sf) const {
+        return value_cast<ValueType>(with_value([&] (const FragmentedView auto& v) { return t.deserialize(v, sf); }));
+    }
+
```

Comment:
This patch does two things: change the data structure, and change the API. This mixes an intense-but-small change with a boring-but-large change, requiring effort from the reviewer. Please split it into a patch that only changes the APIs but without any semantic change, and a patch that changes the data structure to be fragmented (and so all the semantic changes will be concentrated into that patch).

The patch that changes the API should explain the new APIs and the motivation for introducing them.

Reply:
Done.

I would have done that from the start if only splitting header-touching commits wasn't so annoying (because of #1).